src/analysis/visitor-transfer-function.h

```diff
@@ -0,0 +1,108 @@
+#ifndef wasm_analysis_visitor_transfer_function_h
+#define wasm_analysis_visitor_transfer_function_h
+
+#include <queue>
+
+#include "cfg.h"
+#include "lattice.h"
+#include "wasm-traversal.h"
+
+namespace wasm::analysis {
+
+// Utility for visitor-based transfer functions for forward and backward
+// analysis. Forward analysis is chosen by default unless the template parameter
+// Backward is true.
+template<typename SubType, typename Lattice, bool Backward = false>
+struct VisitorTransferFunc : public Visitor<SubType> {
+protected:
+  typename Lattice::Element* currState = nullptr;
+
+  // There are two distinct phases in the execution of the analyzer. First,
+  // the worklist algorithm will be run to obtain a solution, calling
+  // transfer() for each block. Then, to collect the final states, the analyzer
+  // will iterate over every block, calling collectResults().
+
+  // As there is only one set of visitor functions, they are used both to
+  // mutate intermediate states as the transfer function, and also to
+  // collect results from the final states. Therefore, we have the variable
+  // collectingResults to signal whether we are collecting results from the
+  // solved analysis, as opposed to solving the analysis to the visitor
+  // functions.
+  bool collectingResults = false;
+
+public:
+  // Returns an iterable to all the BasicBlocks which depend on currBlock for
+  // information.
+  BasicBlock::BasicBlockIterable getDependents(const BasicBlock* currBlock) {
+    if constexpr (Backward)
+      return currBlock->preds();
+    else
+      return currBlock->succs();
+  }
```

Comment:
This still needs to follow our braces style!
```suggestion
    if constexpr (Backward) {
      return currBlock->preds();
    } else {
      return currBlock->succs();
    }
  }
```